File: leetcode/199-binary-tree-right-side-view.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
  var ret = [];
  for (var stk = [[root, 0]]; stk.length > 0; ) {
    var ele = stk.pop();
    var node = ele[0];
    var level = ele[1];

    if (node === null) continue;
    if (level >= ret.length) ret.push(node.val);
    stk.push([node.left, level + 1]);
    stk.push([node.right, level + 1]);
  }
  return ret;
};

File: leetcode/199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 public:
  vector<int> rightSideView(TreeNode* root) {
    vector<int> ret;
    if (!root) return ret;

    queue<TreeNode*> q;
    for (q.push(root); !q.empty(); ) {
      int n = q.size();
      int val;
      for (int i = 0; i < n; ++i) {
        TreeNode* nd = q.front(); q.pop();
        val = nd->val;
        if (nd->left) q.push(nd->left);
        if (nd->right) q.push(nd->right);
      }
      ret.push_back(val);
    }
    return ret;
  }
};
